fix: increase-camera-quality

EArminjon opened the pull request:
ML Kit need to have image in good quality else it will not be able to process images with precision. Bad precision can lead to bad code recognition and produce wrong detection.

This PR can maybe replace this PR and fix the aspect ratio issue mentioned : https://github.com/juliansteenbakker/mobile_scanner/pull/612
This PR can fix theses issues : https://github.com/juliansteenbakker/mobile_scanner/issues/674, https://github.com/juliansteenbakker/mobile_scanner/issues/752, https://github.com/juliansteenbakker/mobile_scanner/pull/749

I'm not a native expert. I tested and it works well, if you see something weird, please comment.

mocmeow commented:
I've tested your fix and it do really improve the scan distance and speed, difference is noticeably specially on not a really good cam like 5 or 8mpx cam and 16 digits barcode. No problem has been found so far, I'm waiting for a merge soon. 
Great job, thank you!

EArminjon commented:
Hello @juliansteenbakker

What do you think of this PR ?

juliansteenbakker commented on android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScanner.kt:
```diff
@@ -166,6 +170,24 @@ class MobileScanner(
+    private fun getResolution(windowManager: WindowManager): Size {
+        val rotation = windowManager.defaultDisplay.rotation
+        val widthMaxRes = 480 * 4;
+        val heightMaxRes = 640 * 4;
```
Thanks for the update and sorry for the delay. I have just one question: The resolution is now set to 2560 * 1920. Wouldn't it be better to make this user configurable, so in case there is an issue on some device, the resolution can be scaled down. I can imagine some older devices having trouble with this high resolution.

navaronbracke replied:
I agree with this point. The `camera` package does something similar.

EArminjon replied:
Okay, i can let user provide two parameters : maxResolutionHeight and maxResolutionWidth.
Where must i add that ? Inside MobileScannerController properties or inside the MobileScanner properties ?
As it is only used by Android, must i put that inside an AndroidSettings models ?

EArminjon replied:
Also note that, Android said if the resolution is not available, Android will take the closest one.

navaronbracke replied:
@EArminjon The configuration should be passed to the Dart-side and the native side. So you'll have to do some plumbing here and there.

If this is only for Android, then documenting that limitation is good enough.

I would go with a https://api.flutter.dev/flutter/dart-ui/Size-class.html as the type of this parameter (so that we do not have 2 values to pass around)

You can use the newer ResolutionSelector API to make it work on newer version of Android.

EArminjon replied:
Yep i will provide a size, more efficient.
I tried to use the new api but I didn't achieve to make something working. Android native is not really easy for me...

EArminjon commented:
Usage : 

```dart
MobileScannerController(
    androidResolution: const Size(1920, 2560),
)
```

navaronbracke commented on lib/src/mobile_scanner_controller.dart:
```diff
@@ -58,9 +59,25 @@ class MobileScannerController {
   /// Automatically start the mobileScanner on initialization.
   final bool autoStart;
 
+  /// Can be used to override default Android camera resolution.
```
Could you rename this property to just `cameraResolution` ? Maybe one day, we can use this to configure resolutions on other platforms as well.

```suggestion
/// The desired resolution for the camera.
///
/// When this value is provided, the camera will try to match this resolution,
/// or fallback to the closest resolution available.
/// When this is null, Android defaults to a resolution of 640x480.
///
/// Bear in mind that changing the resolution has an effect on the aspect ratio.
///
/// When the camera orientation changes,
/// the resolution will be flipped to match the new dimensions of the display.
///
/// Currently only supported on Android.
final Size? cameraResolution;
```

navaronbracke commented on android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScanner.kt:
```diff
@@ -166,6 +170,24 @@ class MobileScanner(
 
+    // Return the best resolution for the actual device orientation.
+    // By default camera set its resolution to width 480 and height 640 which is too low for ML KIT.
+    // If we return an higher resolution than device can handle, camera package take the most relevant one available.
```
What do you mean with 

```
If we return an higher resolution than device can handle, camera package take the most relevant one available.
```

Is the CameraX library handling it for us already? (Not a CameraX expert myself)

EArminjon replied:
Android camera library will adapt the requested resolution if not possible : 
![image](https://github.com/juliansteenbakker/mobile_scanner/assets/37028599/4495d730-d4a7-46e5-af23-8231801f3761)

navaronbracke commented on android/src/main/kotlin/dev/steenbakker/mobile_scanner/MobileScanner.kt:
```diff
@@ -229,7 +252,22 @@ class MobileScanner(
+                // Override initial resolution
+                analysisBuilder.setTargetResolution(getResolution(windowManager, androidResolution))
+                // Listen future orientation change to apply the custom resolution
+                displayManager.registerDisplayListener(object : DisplayManager.DisplayListener {
```
I have a feeling that this display listener does several things wrong for now.

1. We recreate it every time
2. We never unset the listener when we no longer need it

Could you look into this? If you can't find a solution right away, could you add a TODO in the code and create a Github issue? Then we can follow up with this later.

My first guess would be to set up a display listener as a property of the MobileScanner,
keep track of the analysisBuilder (if possible) and do some delegation with that.

navaronbracke reviewed (changes requested):
Some minor remarks, but otherwise this looks good.

EArminjon commented:
> @EArminjon did you manage to use the newer API, which replaces the deprecated setTargetResolution on newer Android versions? That is going to be a requirement for this as well.
That also mean current camera2 package to camera2 version 1.3. 

I'm not native expert, could take responsability of this :/

navaronbracke commented:
I can take a look for that for you.

navaronbracke commented:
@EArminjon I found a solution for the display listener (and the deprecated window display getter)

EArminjon commented:
@navaronbracke  PR updated : 
- variable renamed from androidResolution to cameraResolution.
- listener is now disposed (ty for pointing this)

navaronbracke commented:
@EArminjon Apparently we did things side-by-side :)

I fixed the display listener issue in a slightly different way in https://github.com/juliansteenbakker/mobile_scanner/pull/785
so that we don't need to keep track of the window manager as well.

I also fixed a small warning about a private method in the Android code and fixed a deprecated API usage.

Your original commits are included in that PR as well (except for the things we did in parallel)

EArminjon commented:
> @EArminjon Apparently we did things side-by-side :)
> 
> I fixed the display listener issue in a slightly different way in #785 so that we don't need to keep track of the window manager as well.
> 
> I also fixed a small warning about a private method in the Android code and fixed a deprecated API usage.
> 
> Your original commits are included in that PR as well (except for the things we did in parallel)

I'm happy to see that my fix was closest to yours, i'm pretty junior in Android Native. 
Ty for making this better ressource cleanup <3 !

I close this PR right ?

navaronbracke commented:
Yes that is correct. While looking at your latest commit I also found an error I made, we can indeed use `DisplayManager.DisplayListener` with an anonymous object (instead of using a separate class)

I added that to the other PR as well. Thank you!
